azkaban-common/src/main/java/azkaban/executor/ExecutorManager.java

```diff
@@ -934,7 +934,9 @@ public void cancelExecutingJobs(ExecutableFlow exFlow, String userId,
   @Override
   public String submitExecutableFlow(ExecutableFlow exflow, String userId)
     throws ExecutorManagerException {
-    synchronized (exflow) {
+
+    String exFlowKey = exflow.getProjectName() + "." + exflow.getId() + ".submitFlow";
+    synchronized (exFlowKey.intern()) {
```

Comment:
@chengren311 
You pretty much summarized my concerns in your above comment. I am good with the `intern()` approach as a hack for something better later on rather than use something that's breakable/buggy. One suggestion is to add your explanation comment in the synchronization logic code. I was thinking of the HashMap lock management approach but I agree it has a significant implementation overhead.